### PR TITLE
Feature: implement list rule names by target

### PR DESCRIPTION
**Pull request by etiago**
## Motivation
I'm raising this PR to increase LocalStack's coverage of AWS APIs. Currently ListRuleNamesByTarget is not implemented.

It fixes issue #12631 .

## Changes
This PR makes it so that instead of returning `NotImplementedError`, LocalStack will start returning the actual list of EventBus rules for a given target.

## Testing
The changes are covered by snapshot/parity tests.

The happy path can also be tested manually by spinning up LocalStack and following these steps:
```bash
awslocal sqs create-queue --queue-name test-sqs-queue
QUEUE_ARN=$(awslocal sqs get-queue-attributes \
  --queue-url http://sqs.eu-west-2.localhost.localstack.cloud:4566/000000000000/test-sqs-queue \
  --attribute-names QueueArn \
  --query 'Attributes.QueueArn' \
  --output text)
awslocal events create-event-bus --name test-event-bus
awslocal events put-rule \
  --name rule-to-sqs \
  --event-bus-name test-event-bus \
  --event-pattern '{"source": ["test-source"]}'
RULE_ARN=$(aws events describe-rule \
  --name rule-to-sqs \
  --event-bus-name test-event-bus \
  --query 'Arn' \
  --output text)
awslocal events put-targets \
  --rule rule-to-sqs \
  --event-bus-name test-event-bus \
  --targets "Id"="1","Arn"="$QUEUE_ARN"

# Verify that the correct target ARN returns the rule:
awslocal events list-rule-names-by-target --target-arn $QUEUE_ARN --event-bus-name test-event-bus

# Verify that a non-existant target ARN returns an empty rule list:
awslocal events list-rule-names-by-target --target-arn "non-existant-arn" --event-bus-name test-event-bus
```

### CODE REVIEW

**Comment by etiago** [comment]
I have read the CLA Document and I hereby sign the CLA

File: tests/aws/services/events/test_events.py [hunk 1]
```diff
@@ -1742,6 +1742,293 @@ def test_process_pattern_to_single_matching_rules_single_target(
...
+                # Ignore if rule or event bus was already deleted
+                pass
+
+        # Delete custom event bus if we created one
+        if bus_name != "default":
+            try:
+                aws_client.events.delete_event_bus(Name=bus_name)
+            except aws_client.events.exceptions.ResourceNotFoundException:
+                # Ignore if event bus was already deleted
+                pass
```
Comment (dominikschubert) [inline tests/aws/services/events/test_events.py:1842]: :bulb: A general pattern we try to follow, is to be very defensive about cleanups to avoid leaking resources. This is particularly important when running tests against AWS, where they might continue incurring cost when leaked.

In that spirit we typically avoid cleanup code like this:

```python
# creating resource_1
# creating dependent resource_2
...
delete_resource_2(...)
delete_resource_1(...)
...
```
even when wrapped in try/catch blocks.

Instead we have a pytest fixture called `cleanups` which is a simple list of callables that are called in reverse order of insertion when the test has finished executing.

Given my example above, this would correspond to:
```python
# creating resource_1
...
cleanups.append(lambda: delete_resource_1(...))
 
...

# creating dependent resource_2
...
cleanups.append(lambda: delete_resource_2(...))
```

File: tests/aws/services/events/test_events.py [hunk 2]
```diff
@@ -1742,6 +1742,293 @@ def test_process_pattern_to_single_matching_rules_single_target(
...
+
+    @markers.aws.validated
+    @pytest.mark.parametrize("bus_name", ["custom", "default"])
+    def test_list_rule_names_by_target_no_matches(
```
Comment (dominikschubert) [inline tests/aws/services/events/test_events.py:1980]: question: Is there a particular reason this deviates from the cleanup structure in your other tests?

File: tests/aws/services/events/test_events.py [hunk 3]
```diff
@@ -1742,6 +1742,293 @@ def test_process_pattern_to_single_matching_rules_single_target(
...
+        )
+        # Store the original NextToken value before replacing it for snapshot comparison
+        original_next_token = response["NextToken"]
+        # Transform NextToken to a fixed value for snapshot comparison
+        if "NextToken" in response:
+            response["NextToken"] = "<next-token>"
+        snapshot.match("second_page", response)
+        all_rule_names.extend(response["RuleNames"])
+        assert len(response["RuleNames"]) == 2
+        assert "NextToken" in response
```
Comment (dominikschubert) [inline tests/aws/services/events/test_events.py:1930]: nit: While asserts are ok to leave in, we typically try to avoid it for testing API responses as it's more brittle than the snapshots and while technically redundant, it does come with an additional maintenance overhead, particularly when rewriting / refactoring a test. For parity tests a good guideline is to keep them mostly for asserting behavior or if you want to be extremely explicit in what edge case the test case is covering.

File: tests/aws/services/events/test_events.py [hunk 4]
```diff
@@ -1742,6 +1742,293 @@ def test_process_pattern_to_single_matching_rules_single_target(
...
+        aws_client,
+        snapshot,
+        clean_up,
+    ):
+        """Test the ListRuleNamesByTarget API with pagination to verify it correctly handles limits and next tokens."""
+        # Create an SQS queue to use as a target
+        queue_name = f"queue-{short_uid()}"
+        queue_url = sqs_create_queue(QueueName=queue_name)
+        queue_arn = sqs_get_queue_arn(queue_url)
+        snapshot.add_transformer(snapshot.transform.regex(queue_arn, "<queue-arn>"))
```
Comment (dominikschubert) [inline tests/aws/services/events/test_events.py:1862]: This doesn't actually do anything here since we're not capturing any response with a queue ARN
```suggestion
```

Similar with the other test cases

File: tests/aws/services/events/test_events.py [hunk 5]
```diff
@@ -1742,6 +1742,293 @@ def test_process_pattern_to_single_matching_rules_single_target(
...
+            rule_names.append(rule_name)
+            events_put_rule(
+                Name=rule_name,
+                EventBusName=bus_name,
+                EventPattern=json.dumps({"source": [f"source-{i}"]}),
+            )
+
+            # Add the SQS queue as a target for this rule
+            target_id = f"target-{i}"
+            snapshot.add_transformer(snapshot.transform.regex(target_id, f"<target-id-{i}>"))
```
Comment (dominikschubert) [inline tests/aws/services/events/test_events.py:1887]: Like with the SQS ARN and EventBus name, we don't capture target ids in the response
```suggestion
```

File: tests/aws/services/events/test_events.py [hunk 6]
```diff
@@ -1742,6 +1742,293 @@ def test_process_pattern_to_single_matching_rules_single_target(
...
+        queue_name = f"queue-{short_uid()}"
+        queue_url = sqs_create_queue(QueueName=queue_name)
+        queue_arn = sqs_get_queue_arn(queue_url)
+        snapshot.add_transformer(snapshot.transform.regex(queue_arn, "<queue-arn>"))
+
+        # Create an event bus if using custom bus
+        if bus_name == "custom":
+            bus_name = f"bus-{short_uid()}"
+            events_create_event_bus(Name=bus_name)
+            snapshot.add_transformer(snapshot.transform.regex(bus_name, "<bus-name>"))
```
Comment (dominikschubert) [inline tests/aws/services/events/test_events.py:1868]:
```suggestion
```
Same case as with the SQS ARN, we're not actually capturing a bus name anywhere in the collected snapshots here.

File: tests/aws/services/events/test_events.py [hunk 7]
```diff
@@ -1742,6 +1742,293 @@ def test_process_pattern_to_single_matching_rules_single_target(
...
+        response = aws_client.events.list_rule_names_by_target(
+            TargetArn=queue_arn,
+            EventBusName=bus_name,
+            Limit=2,
+        )
+        # Store the original NextToken value before replacing it for snapshot comparison
+        original_next_token = response["NextToken"]
+        # Transform NextToken to a fixed value for snapshot comparison
+        if "NextToken" in response:
+            response["NextToken"] = "<next-token>"
```
Comment (dominikschubert) [inline tests/aws/services/events/test_events.py:1908]: Q: Typically we'd solve this with a transformer. Any particular reason to avoid doing that here, as well as in the other cases?

**Review by dominikschubert — CHANGES_REQUESTED** [review]
Hey @etiago ! Great job on the first iteration here and thanks for tackling this. :rocket:
Showing great attention already to cleaning up resources and covering different edge cases :clap:  

There's a few minor nits and things we should address before it being ready to merge, particularly for the tests to make sure they follow our best practices for parity testing.
The comments are not added exhaustively, so please extrapolate accordingly and adapt similar cases to the ones commented one. 

Let me know if you have any questions :+1:

File: tests/aws/services/events/test_events.py (hunk 3 above)
Comment (etiago) [reply tests/aws/services/events/test_events.py:1930]: It's a good point... I just had a habit of asserting at the end of tests, but you're right, since we have the snapshots which do a 1:1 validation, the asserts add no value and I see what you mean about the maintenance cost going forwards.

Will remove these.

File: tests/aws/services/events/test_events.py (hunk 7 above)
Comment (etiago) [reply tests/aws/services/events/test_events.py:1908]: No reason... I was stuck thinking of the `regex` transformer which wouldn't work here since the `NextToken` is different each time.

But since you mentioned this, I now spotted we have a `jsonpath` transformer which makes this much simpler, since I can now just have a single
```python
        snapshot.add_transformer(
            snapshot.transform.jsonpath(
                jsonpath="$..NextToken",
                value_replacement="<next-token>",
                reference_replacement=True,
            )
        )
```

And this replaces all instances of `NextToken` and keeps them numbered when verifying multiple times.

I regenerated the snapshot and it's all good now.

File: tests/aws/services/events/test_events.py (hunk 2 above)
Comment (etiago) [reply tests/aws/services/events/test_events.py:1980]: No reason...

To be completely honest, looks like I "forgot" to do it for this test.

In hindsight, after reading your other comment above about cleanups, I now realise that this test is actually the correct one in that as long as resources are created using the existing fixtures, those resources are already tracked and automatically cleaned up at the end of the test.

File: tests/aws/services/events/test_events.py (hunk 1 above)
Comment (etiago) [reply tests/aws/services/events/test_events.py:1842]: This makes sense.

Having looked through my test again, and also looking through some of the existing tests, I think I don't even need to do this via `cleanups`.

Best as I can tell, looking at the `events_create_event_bus` fixture [here](https://github.com/localstack/localstack/blob/master/localstack-core/localstack/testing/pytest/fixtures.py#L2437) for example, `events_create_event_bus` actually yields a function which when I call it to create event buses, it also adds them to a list to keep track of them. And on tear down, it cycles through that list, cleaning them up.

So I think I can safely remove the whole cleanup code and as long as I use the existing fixtures to create resources, my tests don't need to explicitly worry with cleaningup.

**Review by bblommers — APPROVED** [review]

**Review by dominikschubert — APPROVED** [review]
Thanks for addressing the issues and great work on identifying that fixtures are doing their own cleanup already, reducing the need for manual cleanup!

Congratulations on the first contribution to LocalStack :partying_face:
